### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive for non-null assertion on `unknown`

### ISSUE THREAD

**Issue by matthewvalentine** [typescript-eslint/typescript-eslint#10831]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=MYewdgzgLgBAZiEAuGBXMBrMIDuYYC8MYqANqQNwBQoksARgIYBOKI9AVgKbCxEIgAhNVrQYTAF4oA3gF9C8RMKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const foo: unknown = null;
const bar: object = foo!;
const baz: {} = foo!;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected no error, because the non-null assertion is actually necessary. Typescript will allow you to assign a non-null asserted `unknown` to `object` or `{}`, but won't if it isn't non-null asserted. Because an `unknown` value can be `null | undefined`, whereas `object` and `{}` disallow that.

### Actual Result

There is an error on the non-null assertion. If autofixed, it turns into a Typescript error instead, because the assertion was not unnecessary.
